src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -102,22 +106,39 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
   }
 
   private String getLink(String id, String query) {
-    String cse_request = CSE_URL + "?key=" + API_KEY + "&cx=" + id + "&q=" + query;
-    HttpRequest linkRequest =
-        HttpRequest.newBuilder()
-            .GET()
-            .uri(URI.create(cse_request))
-            .setHeader("User-Agent", "Java 11 HttpClient Bot")
-            .build();
-    HttpResponse<String> linkResponse;
-    try {
-      linkResponse = httpClient.send(linkRequest, HttpResponse.BodyHandlers.ofString());
-    } catch (IOException | InterruptedException ex) {
+    String url = CSE_URL + "?key=" + API_KEY + "&cx=" + id + "&q=" + query;
+    try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+      try (CloseableHttpResponse response = httpClient.execute(new HttpGet(url))) {
+
+        if (response.getStatusLine().getStatusCode() != 200) {
+          return null;
+        }
+        HttpEntity entity = response.getEntity();
+        if (entity == null) {
+          return null;
+        }
+        JsonParser parser = new JsonParser();
+        JsonObject obj;
+        try {
+          obj = parser.parse(new InputStreamReader(entity.getContent())).getAsJsonObject();
+        } catch (JsonIOException
+            | JsonSyntaxException
+            | UnsupportedOperationException
+            | IOException e) {
+          return null;
+        }
+        if (!obj.isJsonNull()
+            && !obj.get(CSE_ITEMS).getAsJsonArray().isJsonNull()
```

Comment:
obj.get can return null if CSE_ITEMS isn't a valid key. Can guard with has()

Reply:
Note that you don't need these isJsonNull() checks. get() will definitely return null if obj was really the JSONObject representing null.

```java
if (obj.has(CSE_ITEMS) && obj.getAsJsonArray(CSE_ITEMS).size() > 0) {
```
is what you're looking for I think. Note how you can straightaway getAsJsonArray(CSE_ITEMS) without doing a get() first.